public/app/features/logs/logsFrame.ts

```diff
@@ -0,0 +1,68 @@
+import { DataFrame, FieldCache, FieldType, FieldWithIndex, DataFrameType } from '@grafana/data';
+
+import { parseLegacyLogsFrame } from './legacyLogsFrame';
+
+// these are like Labels, but their values can be
+// arbitrary structures, not just strings
+export type Attributes = Record<string, unknown>;
+
+// the attributes-access is a little awkward, but it's necessary
+// because there are multiple,very different dataframe-represenations.
+export type LogsFrame = {
+  timeField: FieldWithIndex;
+  bodyField: FieldWithIndex;
+  timeNanosecondField?: FieldWithIndex;
+  severityField?: FieldWithIndex;
+  idField?: FieldWithIndex;
+  attributes?: Attributes[];
+};
+
+function getField(cache: FieldCache, name: string, fieldType: FieldType): FieldWithIndex | undefined {
+  const field = cache.getFieldByName(name);
+  if (field == null) {
+    return undefined;
+  }
```

Comment:
Also return `null` here? What would be the difference to https://github.com/grafana/grafana/pull/70318/files#diff-07da85e1ede94286e302d4831fd8768e312dd25291a32fb9bfb1f02115331121R42-R44?

Reply:
hmm.. for now i changed it to be more regular to:
```js
if (field === undefined) {
    return undefined;
}
```

Reply:
and opene a discussion at the bottom.